app/src/main/java/org/schabi/newpipe/info_list/InfoListAdapter.java

```diff
@@ -330,34 +306,28 @@ public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, int
         }
         if (holder instanceof InfoItemHolder) {
             // If header isn't null, offset the items by -1
-            if (header != null) {
+            if (hasHeader()) {
                 position--;
             }
 
             ((InfoItemHolder) holder).updateFromItem(infoItemList.get(position), recordManager);
-        } else if (holder instanceof HFHolder && position == 0 && header != null) {
-            ((HFHolder) holder).view = header;
-        } else if (holder instanceof HFHolder && position == sizeConsideringHeaderOffset()
-                && footer != null && showFooter) {
-            ((HFHolder) holder).view = footer;
         }
     }
 
     @Override
-    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder, final int position,
+    public void onBindViewHolder(@NonNull final RecyclerView.ViewHolder holder,
+                                 final int position,
                                  @NonNull final List<Object> payloads) {
-        if (!payloads.isEmpty() && holder instanceof InfoItemHolder) {
-            for (final Object payload : payloads) {
-                if (payload instanceof StreamStateEntity) {
-                    ((InfoItemHolder) holder).updateState(infoItemList
-                            .get(header == null ? position : position - 1), recordManager);
-                } else if (payload instanceof Boolean) {
-                    ((InfoItemHolder) holder).updateState(infoItemList
-                            .get(header == null ? position : position - 1), recordManager);
-                }
-            }
-        } else {
+        if (payloads.isEmpty() || !(holder instanceof InfoItemHolder)) {
             onBindViewHolder(holder, position);
+            return;
+        }
+
+        for (final Object payload : payloads) {
+            if (payload instanceof StreamStateEntity || payload instanceof Boolean) {
+                ((InfoItemHolder) holder).updateState(
+                        infoItemList.get(hasHeader() ? position - 1 : position), recordManager);
+            }
```

Comment:
What does this do? Why is `updateState` called multiple times if the payloads are more than one? What are payloads anyway and why are they unused?

Reply:
You might want to ask @nv95 because he implemented the original version of it:
https://github.com/TeamNewPipe/NewPipe/commit/93f25181596792ed35287cd12fff55d97aa6ffe0#diff-d075c1aaf95acf093a7feaa093198bfbc9c4d5b479e2b57effdaf7a9e74448d9R316

I just niceified the code a bit 😄 

I'm not sure if it is ever used in the first place because I can't find a method calling it (maybe some internal calls?).


If I have time I will try to figure out what the snippet does.

Reply:
Usually payloads is just list of one item (or empty), but docs says:
> payloads – A non-null list of merged payloads. Can be empty list if requires full update.


And `updateState` called with differed parameters

Reply:
@nv95 
Thank you for the quick response.

But what does the ``payload instanceof StreamStateEntity || payload instanceof Boolean``-line do/or is for?
Why not always call ``updateState`` on the holder regardless of the payload?

Reply:
According to the documentation of `onBindViewHolder(VH, int, List<Object>)`:
```java
The payloads parameter is a merge list from {@link #notifyItemChanged(int, Object)} or
{@link #notifyItemRangeChanged(int, int, Object)}.  If the payloads list is not empty,
the ViewHolder is currently bound to old data and Adapter may run an efficient partial
update using the payload info.  If the payload is empty,  Adapter must run a full bind.
Adapter should not assume that the payload passed in notify methods will be received by
onBindViewHolder().  For example when the view is not attached to the screen, the
payload in notifyItemChange() will be simply dropped.
```

So the payloads must come from one (or both) of `notifyItemChanged(int, Object)` and `notifyItemRangeChanged(int, int, Object)`. We only use `notifyItemRangeChanged` [in `FeedFragment`](https://github.com/TeamNewPipe/NewPipe/blob/402990dd9d4b58fef26bbe9aa5a6f9c587011342/app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.kt#L555) and a payload is actually being passed: `StreamItem.UPDATE_RELATIVE_TIME`, of type `int` and value `1`. So it would not pass the check `payload instanceof StreamStateEntity || payload instanceof Boolean` (and as expected when resuming the feed fragment the durations are not reloaded, to test this just mark an item as watched and pause&resume the app).

Another note: according to the documentation above, "If the payload is empty,  Adapter must run a full bind.", but this is not being done in our code.

Reply:
> I think this is done:

Oh right, sorry ;-)

Reply:
I think you can remove [that function](https://github.com/TeamNewPipe/NewPipe/blob/cb78433a26989691f4b7b1d30aa25f0c469e8bfb/app/src/main/java/org/schabi/newpipe/info_list/InfoListAdapter.java#L314-L331) completely, and let the base function always do the full update.

The payload `StreamItem.UPDATE_RELATIVE_TIME` is [actually being received](https://github.com/TeamNewPipe/NewPipe/blob/d5199eac3e9e6cf8331adbbe9b742b9b1d178f7e/app/src/main/java/org/schabi/newpipe/local/feed/item/StreamItem.kt#L53-L63), but it would never be received by the method I told you to remove since [`FeedFragment` uses `GroupieAdapter`](https://github.com/TeamNewPipe/NewPipe/blob/402990dd9d4b58fef26bbe9aa5a6f9c587011342/app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.kt#L148-L151) and not `InfoListAdapter`.

Reply:
Done 😄